lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,485 @@
+  SmallVector<Value, 8> lhsValues;
+  getAllBundleLowerings(lhs, lhsValues);
+
+  SmallVector<Value, 8> rhsValues;
+  getAllBundleLowerings(rhs, rhsValues);
+
+  // Check that we got out the same number of values from each bundle.
+  assert(lhsValues.size() == rhsValues.size() &&
+         "connected bundles don't match");
+
+  for (auto tuple : llvm::zip_first(lhsValues, rhsValues)) {
+    Value newLhs = std::get<0>(tuple);
+    Value newRhs = std::get<1>(tuple);
+    builder->create<ConnectOp>(op.getLoc(), newLhs, newRhs);
+  }
+
+  removeOp(op);
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Helpers to manage state.
+//===----------------------------------------------------------------------===//
+
+static DictionaryAttr getArgAttrs(StringAttr nameAttr, StringRef suffix,
+                                  OpBuilder *builder) {
+  SmallString<16> newName(nameAttr.getValue());
+  newName += suffix;
+
+  StringAttr newNameAttr = builder->getStringAttr(newName);
+  Identifier identifier = builder->getIdentifier("firrtl.name");
+  NamedAttribute attr = NamedAttribute(identifier, newNameAttr);
+
+  return builder->getDictionaryAttr(attr);
+}
+
+// Creates and returns a new block argument of the specified type to the module.
+// This also maintains the name attribute for the new argument, possibly with a
+// new suffix appended.
+Value FIRRTLTypesLowering::addArgument(Type type, unsigned oldArgNumber,
+                                       StringRef nameSuffix) {
+  FModuleOp module = getOperation();
+  Block *body = module.getBodyBlock();
+
+  // Append the new argument.
+  auto newValue = body->addArgument(type);
+
+  // Keep the module's type up-to-date.
+  module.setType(builder->getFunctionType(body->getArgumentTypes(), {}));
+
+  // Copy over the name attribute for the new argument.
+  StringAttr nameAttr = getFIRRTLNameAttr(module.getArgAttrs(oldArgNumber));
+  if (nameAttr) {
+    auto newArgAttrs = getArgAttrs(nameAttr, nameSuffix, builder);
+    module.setArgAttrs(newValue.getArgNumber(), newArgAttrs);
+  }
+
+  // Remember to delete the original block argument.
+  removeArg(oldArgNumber);
+
+  return newValue;
+}
+
+// Store the mapping from a bundle typed value to a mapping from its field names
+// to flat values.
+void FIRRTLTypesLowering::setBundleLowering(Value oldValue, StringRef flatField,
+                                            Value newValue) {
+  loweredBundleValues[oldValue][flatField] = newValue;
+}
+
+// For a mapped bundle typed value and a flat subfield name, retrieve and return
+// the flat value if it exists.
+Optional<Value> FIRRTLTypesLowering::getBundleLowering(Value oldValue,
+                                                       StringRef flatField) {
+  if (oldValue && loweredBundleValues.count(oldValue) &&
+      loweredBundleValues[oldValue].count(flatField))
+    return Optional<Value>(loweredBundleValues[oldValue][flatField]);
```

Comment:
This is doing a bunch of redundant map lookups, I'd recommend something like:

```
if (!oldValue) return None;
auto lbvIt = loweredBundleValues(oldValue);
if (lbvIt == loweredBundleValues.end()) return None;
auto fieldIt = lbvIt->find(flatField);
if (fieldIt == lbvIt->end()) return None;
return *fieldIt;

```

Reply:
However, when can this ever fail?

Reply:
The way this works right now, the lookup can fail to find a value when we are flattening subfields of subfields. The internal subfield nodes do not fine a mapped value, only the leaf subfields. However, I think this might change as I incorporate some of your other feedback.

Reply:
After re-working and simplifying things, this lookup cannot, in fact, fail. The pass now asserts this.